### ISSUE TITLE
oxfmt: Support `stdin` and `--stdin-filepath`

### ISSUE THREAD

**Issue by KieranP** [oxc-project/oxc#14720]
Prettier supports stdin and has an option called `--stdin-filepath` that allows prettier to infer the file type. It then returns the formatted output to stdout, instead of writing the changes to the disk.

This is helpful for editors, where you can trigger the formatter command on an unsaved file, the editor will call prettier, passing the current unsaved file contents via stdin, provide the current filename to stdin-filepath, prettier then uses these to format accordingly, the formatted content is returned to stdout, and editor updates the unsaved file.

oxfmt currently does not support this, making it difficult to use in editors. If I have an unsaved file, and trigger oxfmt on the file path, it overwrites the file in disk, and the editors contents remain unchanged, forcing me to reload the file and lose any changes. Obviously not ideal.

An option to pass code via stdin and return the formatted code to stdout for the editor to them use would be helpful.

**Comment by Sysix** [comment]
> making it difficult to use in editors. 

Did you try `oxc_language_server` with the `fmt.experimental = true` option :)? 
https://github.com/oxc-project/oxc/tree/main/crates/oxc_language_server#workspace-options

This should already solve your problems with in memory files.

**Comment by KieranP** [comment]
@Sysix The downside to that is that I can't update oxfmt myself when a new release is made. I need to wait on an upstream update of the LSP and then an update of the Zed extension.

A native stdin option for oxfmt would enable me to use oxfmt in this and other cases where you don't want the source file to be written upon formatting, or want to process it first before writing the file to prevent IO. e.g. the following would only write IO once:

```bash
curl http://example.com/file.js > ofmt > file.js
```

**Comment by david-crespo** [comment]
Editors like Helix want you to plug in to their format on save pipeline through stdin. See `formatter` in the [language config docs](https://docs.helix-editor.com/languages.html). Helix also supports LSP formatting but there are many situations where someone who isn't using the language server might want oxfmt formatting.

In the meantime there's always this workaround, which is still probably faster than prettier.

```sh
function oxfmt-stdin() {
  local ext="${1:-ts}"
  local temp_file="$(mktemp /tmp/oxfmt.${ext})"
 
  cat > "$temp_file"
  bunx oxfmt "$temp_file" > /dev/null
  cat "$temp_file"
  rm "$temp_file"
}
```

```console
$ echo "console.log('hi')" | oxfmt-stdin ts
console.log("hi");
```

**Comment by x-ray-s** [comment]
Does anyone have a resolution to use `oxfmt` in Zed? It does not support `--stdin-filepath`, so it can't easily replace Prettier in Zed.

**Comment by miguno** [comment]
This feature is required to integrate oxfmt with neovim: https://github.com/stevearc/conform.nvim/issues/802

**Comment by leaysgur** [comment]
NOTE: IIUC, `conform.nvim` is usable with `stdin = false` (via tmpfile).

**Comment by miguno** [comment]
> NOTE: IIUC, `conform.nvim` is usable with `stdin = false` (via tmpfile).

Interesting. Any pointers for such a setup? I didn't find any details beyond a brief mention at https://oxc.rs/docs/guide/usage/formatter.

**Comment by leaysgur** [comment]
It's not there.

I simply defined a custom formatter according to the documentation. 😅

https://github.com/stevearc/conform.nvim?tab=readme-ov-file#options

**Comment by miguno** [comment]
For others reading this, here's my current setup for neovim with conform.nvim:

```lua
-- Relevant settings for conform.nvim to use oxfmt for JS/TS files.
-- Details and more customization options at https://github.com/stevearc/conform.nvim?tab=readme-ov-file#options
      formatters = {
        oxfmt = {
          command = "oxfmt",
          args = { "$FILENAME" },
          stdin = false,
          -- When stdin=false, use this template to generate the temporary file that gets formatted
          tmpfile_format = ".conform.$RANDOM.$FILENAME",
        },
      },
      formatters_by_ft = {
        -- Use oxfmt as preferred formatter if available,
        -- otherwise use prettier.
        javascript = { "oxfmt", "prettier" },
        javascriptreact = { "oxfmt", "prettier" },
        typescript = { "oxfmt", "prettier" },
        typescriptreact = { "oxfmt", "prettier" },
      },
```

**Comment by typedrifter** [comment]
Thanks @miguno I use it as well as for oxlint --fix, as a solution to my discussion: https://github.com/oxc-project/oxc/discussions/16494

```lua
formatters = {
	oxlint = function()
		local util = require("conform.util")
		return {
			command = util.from_node_modules("oxlint"),
			args = { "$FILENAME", "--fix" },
			stdin = false,
			-- When stdin=false, use this template to generate the temporary file that gets formatted
			tmpfile_format = ".conform.$RANDOM.$FILENAME",
		}
	end,
},
```

**Comment by camc314** [comment]
@leaysgur can this be closed since #16868 has been merged?

**Comment by leaysgur** [comment]
Yes, thank you~ 🙏🏻

**Comment by typedrifter** [comment]
I submitted a PR to `conform.nvim` to add support for this new argument: https://github.com/stevearc/conform.nvim/pull/829. 

Alternatively, `Oxfmt` can also be used as a Language Server, allowing `vim.lsp.buf.format` to trigger it. To enable this, I also made a PR for `nvim-lspconfig`: https://github.com/neovim/nvim-lspconfig/pull/4242.
